rust/connlib/tunnel/src/device_channel/tun_android.rs

```diff
@@ -99,11 +99,11 @@ impl ioctl::Request<GetInterfaceNamePayload> {
         }
     }
 
-    fn name(&self) -> std::borrow::Cow<'_, str> {
+    fn name(&self) -> String {
         // Safety: The memory of `self.name` is always initialized.
         let cstr = unsafe { std::ffi::CStr::from_ptr(self.name.as_ptr() as _) };
 
-        cstr.to_string_lossy()
+        cstr.to_string_lossy().into_owned()
```

Comment:
@conectado This is what fixed the segfault.

cc @thomaseizinger -- this was causing a segfault in some versions of android and only on physical hardware. What's weirder, if I logged a single trace message with `tracing::trace!("test")` inside this function, the segfault would disappear.

One pattern I noticed is that this is occurring on devices that load the 32-bit libs (some 64-bit devices on older Androids do this, e.g. my Chrome book from 2015).

I thought maybe it was all of these things before finally concluding this was causing the issue:

- Thought maybe `fd` was somehow getting dropped
- Thought maybe it was a race condition / timing issue, added delays and still occurred
- Tried replacing this with our old implementation from https://github.com/firezone/firezone/commit/0de16d3676eb24499d329d090926dc87b271e56b (this worked, a clue)
- Tried removing this altogether -- we only use the interface name to read the MTU, which is always 1280.

Reply:
Firebase logs are here (no symbols)

https://console.firebase.google.com/u/0/project/firezone-55040/crashlytics/app/android:dev.firezone.android/issues/efa47ac9c42ab1fb686a258a8b0136a6?time=last-seven-days&types=crash&sessionEventKey=65c60130020500016162e490dfdd5b47_1911995967545350167

Reply:
Scratch that. This fixed the issue on the OnePlus One, but the issue still occurs on the Chromebook with Android 9.

Reply:
I'm actually baffled by this, it seems like we follow all the safety detailed in `Cstr::from_ptr`, the most important one that this could change is that the underlying memory for the CStr needs not to be mutated for as long as it lives but since we're holding an `&self` to the structure that holds the buffer there's no way we could be mutating it.

Also, as soon as this function returns we convert this to string(which we should remove as part of this PR since it's redundant) so we only use it as an owned value.

I believe we're invoking UB at another part of this file which is just bubbling up here.

I think we should take the time to run [miri](https://github.com/rust-lang/miri) with android(ideally, also as part of our CI) as soon as we can because this will probably bite us again in the future.

Reply:
@conectado Yeah, I think you're probably right.

One thing that's consistent is that this only happens in the 32-bit versions of `libconnlib.so`. And I'm about 90% sure it only happens in the `interface_name()` function, but I need to test more to be sure.

I'll take a look at miri. In the meantime, if there's any way to get an `arm` image loaded into the emulator, that might allow us to reproduce it in the simulator at least.

Reply:
@jamilbk we can test something in the meantime(should be equivalent to this function but requires no unsafe):

```rs
fn name(&self) -> String {
  String::from_utf8_lossy(self.name().split(|&c| c == b'\0').next().unwrap_or_default());
}
```

Reply:
@conectado It's a miracle. That fixed it! Seems like a pointer misalignment problem then on some architectures?

Reply:
Either that, or on Android the interface name is not null-terminated?

Reply:
> Either that, or on Android the interface name is not null-terminated?
It should be nul-terminated but..

Maybe, just maybe, the problem was that `CString` normally expects a [single nul-byte](https://doc.rust-lang.org/std/ffi/struct.CString.html#safety) at the end not a sequence, past time I mistakenly used a sequence of nul-bytes at the end I got an error but I wasn't using `from_ptr`.

Of course, this would make the API for `from_ptr` really unintuitive since you can't just directly pass pointers for buffers obtained by calling ffi, in which case I'd question the usefulness of it existing at all.

Reply:
@conectado Hmmmm, scratch that again.

So now the session activity loads, the Resources list is populated, but there's no tun interface visible from my bash shell on the device and packets aren't flowing. I wonder if the interface is actually being created, or if it's being set up properly...

That would explain why trying to retrieve its name fails, and why possibly getting its MTU is failing and preventing packets from flowing.

I'll have to do more testing.

Reply:
Hmm, it's working reliably on the OnePlus one, but not the Chromebook. I'll try a reboot and see if that clears anything out. It's possible the network state is wonky after all the crashes.

Reply:
weird that the `ioctl` is not just failing if the interface doesn't exists

Reply:
Think I found the issue. See #3619

Reply:
This has been tested to work on two devices. The Chromebook is experiencing this, but it's a separate issue related to ChromeOS: https://github.com/firezone/firezone/issues/3620